### ISSUE TITLE
Add safety layer to iterators on failure

### ISSUE THREAD

**Issue by gilrrei** [queens-py/queens#64]
> This issue was originally created by @gilrrei on [gitlab.lrz.de](https://gitlab.lrz.de/queens_community/queens/-/issues/512) on 2024-08-01.

## Motivation and Context
Currently, if the iterator fails during a run for some reason, no current data is exported. As an example, during MCMC for some reason the sampling fails, no samples are stored (which after 500000 samples is really frustrating). We should add a safety to export as much as possible on failure, which is also useful for debugging what went wrong.

This is also an important step towards introducing sensible restarts [46](https://gitlab.lrz.de/queens_community/queens/-/issues/46) 

## Current Behavior
In the VI iterators this is already done by regularly exporting the results.

## Possible Solution
- [ ] first: unify result writing [294](https://gitlab.lrz.de/queens_community/queens/-/issues/294) 
- [ ] second find a solution for this

## Related Issues
* Blocks
* Is blocked by
* Follows
* Precedes
* Related to
* Part of
* Composed of

## Interested Parties
